### PR TITLE
Enable credential provider authentication for self-hosted mysql DB

### PR DESCRIPTION
**Pull request by ssrangisetti**

### CODE REVIEW

File: lib/config/fileconf.go [hunk 1]
```diff
@@ -1540,6 +1540,20 @@ type Databases struct {
 	AWSMatchers []AWSMatcher `yaml:"aws,omitempty"`
 	// AzureMatchers match Azure hosted databases.
 	AzureMatchers []AzureMatcher `yaml:"azure,omitempty"`
+	// TokenConfig has http and authentication config for token exchange service
+	TokenConfig TokenConfig `yaml:"token,omitempty"`
+}
+
+type TokenConfig struct {
```
Comment (amanhigh) [inline lib/config/fileconf.go]: Can we have defaults so client can get started with minimal config. Eg. https://github.com/caarlos0/env or VIPER

File: lib/config/fileconf.go [hunk 2]
```diff
@@ -1540,6 +1540,20 @@ type Databases struct {
...
+	Enabled           types.BoolOption `yaml:"enabled"`
+	UrlTmpl           string           `yaml:"url_tmpl"`
```
Comment (amanhigh) [inline lib/config/fileconf.go]: Can use full names which are self explanatory. Also TokenSourceConfig or something else ?

File: lib/token/auth.go [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+type AuthCreator func(config AuthConfig) AuthInjector
+
+var AuthCreators = make(map[string]AuthCreator)
+
+func init() {
+	AuthCreators[NoneAuthScheme] = newNoAuthInjector
+}
+
+type AuthInjector interface {
```
Comment (amanhigh) [inline lib/token/auth.go]: Can we vet interface against popular cloud Identity services eg. GCP, Azure etc

File: lib/token/client.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+	Password string `json:"password"`
+}
+
+type urlTmplParams struct {
+	DBName   string
+	Token    string
+	Username string
+}
+
+func NewClient(config ClientConfig) *Client {
```
Comment (amanhigh) [inline lib/token/client.go]: Can use a popular client like https://github.com/go-resty/resty

File: lib/token/client_test.go [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+	"fmt"
+	"github.com/stretchr/testify/require"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"text/template"
+	"time"
+)
+
+func TestNewClient(t *testing.T) {
```
Comment (amanhigh) [inline lib/token/client_test.go]: Can avoid test if we use existing client.
https://github.com/go-resty/resty

File: lib/config/configuration_test.go [hunk 6]
```diff
@@ -3403,3 +3404,64 @@ func TestApplyDiscoveryConfig(t *testing.T) {
 		})
 	}
 }
+
+func TestApplyFileConfig_DbTokenAuth_errors(t *testing.T) {
```
Comment (amanhigh) [inline lib/config/configuration_test.go:3407]: Can add tests for missing mandatory fields.

File: lib/config/configuration_test.go [hunk 7]
```diff
@@ -3403,3 +3404,64 @@ func TestApplyDiscoveryConfig(t *testing.T) {
...
+			}, defaultCfg)
+			require.ErrorContains(t, err, test.error)
+		})
+	}
+}
+
+func TestApplyFileConfig_DbTokenAuth(t *testing.T) {
+	defaultCfg := service.MakeDefaultConfig()
+	err := ApplyFileConfig(&FileConfig{
+		Databases: Databases{
```
Comment (amanhigh) [inline lib/config/configuration_test.go:3458]: Can add happy test for Full Config (With Optionals like timeout etc) , Minimal Config (Bare minimum required).

**Review by amanhigh — COMMENTED** [review]
Few other observations which may help user by time feature is compleete
- Diagram explaning interaction with Tokensource in Readme
- Easy setup in local and trial with dummy token source.

File: lib/tokensource/client.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+	return &Client{
+		urlTemplate:  config.UrlTemplate,
+		httpClient:   client,
+		authInjector: authInjector,
+	}
+}
+
+func (c *Client) GetCredentials(ctx context.Context, dbName, username, token string) (string, string, error) {
+	buf := &bytes.Buffer{}
+	if err := c.urlTemplate.Execute(buf, urlTmplParams{DBName: dbName, Token: token, Username: username}); err != nil {
```
Comment (amanhigh) [inline lib/tokensource/client.go]: Optional: move Url Subtitution to buildUrl Helper Function

File: lib/tokensource/client.go [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+	if resp.IsSuccess() {
+		credentialResponse := resp.Result().(*CredentialResponse)
+		return credentialResponse.Response.Username, credentialResponse.Response.Password, nil
+	} else {
+		errorResp := resp.Error().(*CredentialResponse)
+		return "", "", trace.Errorf("%v: %v", errorResp.Code, errorResp.Msg)
+	}
+}
+
+func (c *ClientConfig) CheckAndSetDefaults() {
```
Comment (amanhigh) [inline lib/tokensource/client.go:104]: Should this be called in Constructor NewClient so no one missises calling it ?

File: lib/config/configuration.go [hunk 10]
```diff
@@ -1296,21 +1297,24 @@ func applyKubeConfig(fc *FileConfig, cfg *service.Config) error {
 // applyDatabasesConfig applies file configuration for the "db_service" section.
 func applyDatabasesConfig(fc *FileConfig, cfg *service.Config) error {
 	cfg.Databases.Enabled = true
-	tokenConfig := fc.Databases.TokenConfig
+	tokenConfig := fc.Databases.TokenSourceConfig
 	if tokenConfig.Enabled.Value {
 		var err error
-		cfg.Databases.TokenConfig.UrlTmpl, err = template.New("url").Parse(tokenConfig.UrlTmpl)
+		cfg.Databases.TokenSourceConfig.Enabled = tokenConfig.Enabled.Value
+		if tokenConfig.UrlTemplate == "" {
```
Comment (amanhigh) [inline lib/config/configuration.go:1304]: Can some of these checks be moved to CheckAndSetDefault for encapsulation ?

**Review by amanhigh — APPROVED** [review]
Overall Looks Good Few Comments.

**Review by amanhigh — APPROVED** [review]
Looks Good
